### ISSUE TITLE
Support multiple URLs

### ISSUE THREAD

**Issue by anovinger** [keycloak/keycloak#16432]
### Description

An instance of KeyCloak should be able to support multiple URLs to a Realm to different clients. This used to work in Wildfly but not with Quarkus.

We have subdomain1.domain.com pointing to client 1, and subdomain2.domain.com point to client 2, and we have hostname=subdomain1.domain.com set in the keycloak.conf file. With this setup when we hit subdomain1, it works as expected, when we hit subdomain2, we are redirected to subdomain1. When we updated the hostname config to subdomain2, we have the same results that subdomain2 works correctly, but subdomain1 redirects back to subdomain2.

We are using this setup so we can have employees use one url, and non employees use the other url. We need this setup so our backend Ping know where to look for the user.

### Details

Seems to be someone else have the same issue I am having: https://stackoverflow.com/questions/62249597/keycloak-server-exposed-via-2-public-urls which lead me to https://issues.redhat.com/browse/KEYCLOAK-15553

**Comment by priyankyshah** [comment]
Any update on this one?

**Comment by dmaphy** [comment]
Is there any update on this? Since updating to 26.0.x we are forced to use the hostname.v2 feature and are not able to set the `KC_HOSTNAME_STRICT` to `false` while also having `KC_HOSTNAME_ADMIN` set. It would be neat if it would be possible to define a list as hostname instead of one or have the additional availability to add some kind of hostname aliases (maybe `--hostname-aliases`, `KC_HOSTNAME_ALIASES`?)

**Comment by shawkins** [comment]
> Is there any update on this? 

The original request is for a URL per client. That is not supported outside being able to set hostname-strict to false.

As you point out you cannot use hostname-admin with hostname-strict=false. 

Instead of setting hostname-admin, you could use the proxy to rewrite URLS as needed. That is rewrite hostname1/admin, hostname2/admin, etc. to be admin-hostname/admin. The only thing that won't work well is that once requests are hitting admin-hostname, links to the frontend will remain on admin-hostname - unless they are links specifically to a realm and the realm has a hostname set. So you may want to also rewrite anything that looks like admin-hostname/not-admin to the default frontend url of your choosing. It would need broader discussion to support behavior like this directly in Keycloak. cc @vmuzikar

**Comment by dmaphy** [comment]
I think my issue is basically the same. What I found at my customers site is a blue-green deployment where there are two Keycloak installations which have one of the following set as `hostname`:

* blue-something.com
* green-something.com

We want to have something.com available on one both the installations as well. If we do that and call something.com in the browser, I get redirected to green-something.com, but the `green-` should not be visible to the customer at all. If I set the hostname to something.com and call admin.green.something.com I will get redirected to something.com the other way around. I think this matches quite well to the original description of this issue.

My point was, that it was possible to have the Keycloak installations "listening" on different hostnames when setting `--hostname-strict` to `false`, because then the client "decides" which name he is sending via the HTTP Host header. Even if I understand we don't want this for security purposes and that's maybe one of the reasons why this changed with hostname.v2 my customer was making heavy use of this feature until Keycloak 25. :)

We will have a look if we are able to implement a workaround for this into the reverse proxy instead, but I still think that a kind of hostname alias would be a better solution for this.

**Comment by shawkins** [comment]
Let me see if I can summarize this so that we can close or further refine.

There are users who want to use hostname-strict=false with hostname-admin (or with enabling the dynamic backchannel requests).

The core of the problem is that keycloak in a few places needs to know how to traverse from the front-end to the admin and vice-versa. 

For example when you are in the admin console and click "Manage Account" - that needs to take you to the front-end link for the current realm. If the front end is dynamic, then we don't know where this link should point to.

Thus keycloak has required that if you specify hostname-admin, then you must also provide a full hostname url.

It was described in https://github.com/keycloak/keycloak/issues/16432#issuecomment-2593749835 that you can effectively take responsibility whatever you want your admin access to be in the proxy - you are already doing some heavy lift in the proxy due to using hostname-strict=false, so this isn't entirely unreasonable.

The idea of introducing something like hostname-aliases would be a varition of hostname-strict=false that supplies Keycloak with a list of allowed hostnames - however that would not resolve the issue with hostname-admin and cross navigating. This could be opened in a separate issue / discussion as a way to make hostname-strict=false more secure.

The description of the blue green problem seems to still be a case of hostname-strict=false, or was that to highlight something else?

@dmaphy @anovinger I believe what we'd need to see to move this forward is a proposal that fully defines how you can determine the front-end url when you are accessing keycloak over the admin url.

I'll set this to auto-expire so that if the discussion does not proceed, and we can assume just using hostname-strict=false is good enough for now.

**Comment by dmaphy** [comment]
> The description of the blue green problem seems to still be a case of hostname-strict=false, or was that to highlight something else?

Actually we made our setup working exactly like that, yes. Set `hostname-strict=false` and don't set a `hostname` or `hostname-admin` at all. If that's the way you would recommend I'm fine with that one. What's a bit misleading here is the documentation at https://www.keycloak.org/server/all-config?q=hostname-strict which says:

> Should always be set to true in production, unless your reverse proxy overwrites the Host header. If enabled, the hostname option needs to be specified.

Trying to achieve setting this to `true` is just not possible then for us.

**Comment by shawkins** [comment]
> Trying to achieve setting this to true is just not possible then for us.

Agreed, but from keycloak's perspective the behavior is much more well-defined / secure when using hostname-strict=true. 

I see two possible enhancements that could refine the hostname-strict=false behavior:

1. Something like hostname-aliases in conjunction with hostname-strict=false. However that would be applicable to backchannel requests as well - otherwise those requests won't know what the frontend url should be. This means that true backchannel requests (not via the proxy) would need the host header set to one of the allowed aliases (might be TLS considerations as well).

2. If becomes even more complicated, but you could try to support hostnam-admin as well with wildcard config options (where the syntax is TBD):
hostname-strict=false
hostname-<incoming>=outgoing hostname
hostname-admin-<incoming>=outgoing hostname
...

The same rule as with hostname-admin would apply - if hostname-admin-x is specified, then hostname-x would need to be specified as a URL as well. Also the caveat about backchannel requests from 1 would apply here too.

Other thoughts: 
- Setting a realm specific front-end url seems like it won't be desirable in these scenarios.
- These scenarios wouldn't work well with TLS passthrough, unless a wildcard cert is applicable https://github.com/keycloak/keycloak/issues/21871

cc @vmuzikar
